pkg/provider/chef/chef.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	providerchef.clientName = chefProvider.UserName
+	providerchef.databagService = client.DataBags
+	providerchef.userService = client.Users
+	return providerchef, nil
+}
+
+// Close closes the client connection.
+func (providerchef *Providerchef) Close(_ context.Context) error {
+	return nil
+}
+
+// Validate checks if the client is configured correctly
+// to be able to retrieve secrets from the provider.
+func (providerchef *Providerchef) Validate() (v1beta1.ValidationResult, error) {
+	_, err := providerchef.userService.Get(providerchef.clientName)
+	metrics.ObserveAPICall(ProviderChef, CallChefGetUser, err)
+	if err != nil {
+		return v1beta1.ValidationResultError, fmt.Errorf(errStoreValidateFailed)
+	}
+	return v1beta1.ValidationResultReady, nil
+}
+
+// GetAllSecrets Retrieves a map[string][]byte with the Databag names as key and the Databag's Items as secrets.
+func (providerchef *Providerchef) GetAllSecrets(_ context.Context, _ v1beta1.ExternalSecretFind) (map[string][]byte, error) {
+	return nil, fmt.Errorf("dataFrom.find not suppported")
+}
+
+// GetSecret returns a databagItem present in the databag. format example: databagName/databagItemName.
+func (providerchef *Providerchef) GetSecret(ctx context.Context, ref v1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	if utils.IsNil(providerchef.databagService) {
+		return nil, fmt.Errorf(errUninitalizedChefProvider)
+	}
+
+	key := ref.Key
+	databagName := ""
+	databagItem := ""
+	nameSplitted := strings.Split(key, "/")
+	if len(nameSplitted) > 1 {
+		databagName = nameSplitted[0]
+		databagItem = nameSplitted[1]
+	}
+	log.Info("fetching secret value", "databag Name:", databagName, "databag Item:", databagItem)
+	if databagName != "" && databagItem != "" {
+		return getSingleDatabagItemWithContext(ctx, providerchef, databagName, databagItem, ref.Property)
+	}
+
+	return nil, fmt.Errorf(errInvalidFormat)
+}
+
+func getSingleDatabagItemWithContext(ctx context.Context, providerchef *Providerchef, dataBagName, databagItemName, propertyName string) ([]byte, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+	resultChan := make(chan []byte, 1)
+	errChan := make(chan error, 1)
+
+	go func() {
+		ditem, err := providerchef.databagService.GetItem(dataBagName, databagItemName)
+		if err != nil {
+			errChan <- fmt.Errorf(errNoDatabagItemFound, databagItemName, dataBagName)
+			return
+		}
+
+		jsonByte, err := json.Marshal(ditem)
+		if err != nil {
+			errChan <- fmt.Errorf(errUnableToConvertToJSON)
+			return
+		}
+
+		if propertyName != "" {
+			propertyValue, err := getPropertyFromDatabagItem(jsonByte, propertyName)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			resultChan <- propertyValue
+		} else {
+			resultChan <- jsonByte
+		}
+	}()
+
+	select {
+	case <-ctxWithTimeout.Done():
+		return nil, ctxWithTimeout.Err()
+	case err := <-errChan:
+		return nil, err
+	case result := <-resultChan:
+		return result, nil
+	}
+}
```

Comment:
This function leaks these two channels right now even if the go routine isn't hanging around, the channels are.

Might I suggest an alternative that manages the channel lifecycle inside a function? It's still a buffered channel to avoid having to deal with cancellation inside the sender function, and we can be sure that there is either an error or a result anyways after a while.

And the select will block on the return value. WDYT?

```suggestion
func getSingleDatabagItemWithContext(ctx context.Context, providerchef *Providerchef, dataBagName, databagItemName, propertyName string) ([]byte, error) {
	ctxWithTimeout, cancel := context.WithTimeout(ctx, contextTimeout)
	defer cancel()

	type result = struct {
		values []byte
		err    error
	}

	getWithTimeout := func() chan result {
		resultChan := make(chan result, 1)

		go func() {
			defer close(resultChan)

			ditem, err := providerchef.databagService.GetItem(dataBagName, databagItemName)
			if err != nil {
				resultChan <- result{
					err: fmt.Errorf(errNoDatabagItemFound, databagItemName, dataBagName),
				}

				return
			}

			jsonByte, err := json.Marshal(ditem)
			if err != nil {
				resultChan <- result{
					err: fmt.Errorf(errUnableToConvertToJSON),
				}

				return
			}

			if propertyName != "" {
				propertyValue, err := getPropertyFromDatabagItem(jsonByte, propertyName)
				if err != nil {
					resultChan <- result{err: err}

					return
				}

				resultChan <- result{values: propertyValue}
			} else {
				resultChan <- result{values: jsonByte}
			}
		}()

		return resultChan
	}

	select {
	case <-ctxWithTimeout.Done():
		return nil, ctxWithTimeout.Err()
	case r := <-getWithTimeout():
		if r.err != nil {
			return nil, r.err
		}

		return r.values, nil
	}
}

```